src/main/java/org/broadinstitute/hellbender/engine/ReadsDataSource.java

```diff
@@ -443,12 +443,32 @@ public SAMFileHeader getHeader() {
      */
     private SamFileHeaderMerger createHeaderMerger() {
         List<SAMFileHeader> headers = new ArrayList<>(readers.size());
+        final Set<SAMFileHeader.SortOrder> sortOrders = new HashSet<>(SAMFileHeader.SortOrder.values().length);
```

Comment:
Instead of keeping a separate list of sort orders here, lets add a new method that encapsulates the new logic your adding.  Then just pass the existing list of headers into the new method.
```
SortOrder identifySortOrder(List<SamFileHeader> headers){...} 
```
I think it will make things more readable, and make it easy to test the new functionality.  (If you don't mind could you add a unit test for the new method as well? )